modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v3_2_4/UpgradeDDMContent.java

```diff
@@ -0,0 +1,207 @@
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.dao.jdbc.AutoBatchPreparedStatementUtil;
+import com.liferay.portal.kernel.json.JSONArray;
+import com.liferay.portal.kernel.json.JSONFactory;
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.LocaleUtil;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+import java.text.DecimalFormat;
+import java.text.ParseException;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * @author Rodrigo Paulino
+ */
+public class UpgradeDDMContent extends UpgradeProcess {
+
+	public UpgradeDDMContent(JSONFactory jsonFactory) {
+		_jsonFactory = jsonFactory;
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		try (PreparedStatement ps1 = connection.prepareStatement(
+				StringBundler.concat(
+					"select contentId, data_, DDMStructureVersion.definition ",
+					"from DDMContent join DDMFormInstanceRecordVersion on  ",
+					"storageId = contentId join DDMFormInstanceVersion on ",
+					"DDMFormInstanceVersion.formInstanceId = ",
+					"DDMFormInstanceRecordVersion.formInstanceId and ",
+					"DDMFormInstanceVersion.version = ",
+					"DDMFormInstanceRecordVersion.formInstanceVersion join ",
+					"DDMStructureVersion on DDMStructureVersion.",
+					"structureVersionId = DDMFormInstanceVersion.",
+					"structureVersionId"));
+			PreparedStatement ps2 =
+				AutoBatchPreparedStatementUtil.concurrentAutoBatch(
+					connection,
+					"update DDMContent set data_ = ? where contentId = ?")) {
+
+			try (ResultSet rs = ps1.executeQuery()) {
+				while (rs.next()) {
+					String definition = rs.getString("definition");
+
+					JSONObject definitionJSONObject =
+						_jsonFactory.createJSONObject(definition);
+
+					String data = rs.getString("data_");
+
+					JSONObject dataJSONObject = _jsonFactory.createJSONObject(
+						data);
+
+					if (_upgradeDDMContentData(
+							dataJSONObject.getJSONArray("fieldValues"),
+							definitionJSONObject.getJSONArray("fields"))) {
+
+						ps2.setString(1, dataJSONObject.toJSONString());
+
+						long contentId = rs.getLong("contentId");
+
+						ps2.setLong(2, contentId);
+
+						ps2.addBatch();
+					}
+				}
+			}
+
+			ps2.executeBatch();
+		}
+	}
+
+	private DecimalFormat _getDecimalFormat(Locale locale) {
+		DecimalFormat decimalFormat = _decimalFormatsMap.get(locale);
+
+		if (decimalFormat == null) {
+			decimalFormat = (DecimalFormat)DecimalFormat.getInstance(locale);
+
+			decimalFormat.setGroupingUsed(false);
+			decimalFormat.setMaximumFractionDigits(Integer.MAX_VALUE);
```

Comment:
I just followed the service's implementation. You can check it [here](https://github.com/liferay/liferay-portal/blob/d8129c6aa4b153c4fd7569ee080bf1612c977bb6/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/java/com/liferay/dynamic/data/mapping/form/field/type/internal/numeric/NumericDDMFormFieldUtil.java#L29-L41). Also, I think that users could have submitted decimal numbers with more than 3 digits since the application allows it.